File: src/qsql.js
// ============================================================
// qsql.js — QSQL: Per-predicate typed SQLite adapter for persist
//
// Zero-impedance bridge: Prolog terms → per-predicate SQLite
// tables with typed argument columns.
//
//   price(aapl, 187.68)  →  table "q$price$2"
//                             _key TEXT PRIMARY KEY
//                             arg0 TEXT   = 'aapl'
//                             arg1 REAL   = 187.68
//
// Atoms → TEXT, numbers → REAL/INTEGER, compounds → JSON TEXT.
// SQLite can index and range-scan individual arguments.
//
// Drop-in replacement for persist-sqlite.js:
//   persist(engine, qsqlAdapter(db));
//
// With QJSON codec for BigNum preservation:
//   persist(engine, qsqlAdapter(db, { parse: qjson_parse }));
//
// Adapter interface:
//   setup, insert(key,functor,arity), remove(key), all(predicates),
//   commit, close
//
// Portable: ES5 style (var, function, no arrows).
// ============================================================

// ── Helpers ──────────────────────────────────────────────────

// Sanitize functor name for use in SQL identifiers
function _qsql_safeName(name) {
  return name.replace(/[^a-zA-Z0-9_]/g, "_");
}

// Table name for a predicate: q$functor$arity
function _qsql_tableName(functor, arity) {
  return "q$" + _qsql_safeName(functor) + "$" + arity;
}

// Convert a serialized term arg to a native SQLite value
//   atom   → TEXT (the name)
//   num    → REAL/INTEGER (or TEXT for BigInt/BigDecimal/BigFloat)
//   other  → JSON TEXT
function _qsql_argVal(arg) {
  if (!arg) return null;
  if (arg.t === "a") return arg.n;
  if (arg.t === "n") {
    var v = arg.v;
    if (typeof v === "number") return v;
    return String(v);
  }
  return JSON.stringify(arg);
}

// Extract the QJSON repr from a serialized num arg (if present)
function _qsql_argRepr(arg) {
  if (!arg || arg.t !== "n") return null;
  return arg.r || null;
}

// ── Adapter Factory ──────────────────────────────────────────

function qsqlAdapter(db, options) {
  var _parse = (options && options.parse) || JSON.parse;
  var _known = {};    // "functor/arity" → true
  var _cache = {};    // sql string → prepared statement

  function _stmt(sql) {
    if (!_cache[sql]) _cache[sql] = db.prepare(sql);
    return _cache[sql];
  }

  function _ensureTable(functor, arity) {
    var fa = functor + "/" + arity;
    if (_known[fa]) return;

    var tbl = _qsql_tableName(functor, arity);
    var ddl = 'CREATE TABLE IF NOT EXISTS "' + tbl + '" (_key TEXT PRIMARY KEY';
    for (var i = 0; i < arity; i++) ddl += ", arg" + i;
    ddl += ")";
    db.exec(ddl);

    for (var i = 0; i < arity; i++) {
      db.exec('CREATE INDEX IF NOT EXISTS "ix$' + tbl + '$' + i +
              '" ON "' + tbl + '"(arg' + i + ')');
    }

    _stmt("INSERT OR IGNORE INTO qsql_meta VALUES (?, ?)").run(functor, arity);
    _known[fa] = true;
  }

  return {
    setup: function() {
      db.exec(
        "CREATE TABLE IF NOT EXISTS qsql_meta " +
        "(functor TEXT, arity INTEGER, PRIMARY KEY(functor, arity))"
      );
      var metas = _stmt("SELECT functor, arity FROM qsql_meta").all();
      for (var i = 0; i < metas.length; i++) {
        _known[metas[i].functor + "/" + metas[i].arity] = true;
      }
    },

    insert: function(key, functor, arity) {
      if (functor == null) return;
      _ensureTable(functor, arity);

      var obj = _parse(key);
      var values = [key];
      if (obj.t === "c" && obj.a) {
        for (var i = 0; i < arity; i++) {
          values.push(i < obj.a.length ? _qsql_argVal(obj.a[i]) : null);
        }
      }

      var tbl = _qsql_tableName(functor, arity);
      var ph = "?";
      for (var i = 1; i < values.length; i++) ph += ", ?";
      var sql = 'INSERT OR IGNORE INTO "' + tbl + '" VALUES (' + ph + ')';
      var s = _stmt(sql);
      s.run.apply(s, values);
    },

    remove: function(key) {
      var obj;
      try { obj = _parse(key); } catch(e) { return; }
      var functor, arity;
      if (obj.t === "c") { functor = obj.f; arity = (obj.a || []).length; }
      else if (obj.t === "a") { functor = obj.n; arity = 0; }
      else return;

      var fa = functor + "/" + arity;
      if (!_known[fa]) return;
      var tbl = _qsql_tableName(functor, arity);
      _stmt('DELETE FROM "' + tbl + '" WHERE _key = ?').run(key);
    },

    all: function(predicates) {
      var results = [];
      var metas;
      if (predicates) {
        metas = [];
        var keys = Object.keys(predicates);
        for (var i = 0; i < keys.length; i++) {
          var parts = keys[i].split("/");
          metas.push({ functor: parts[0], arity: parseInt(parts[1], 10) });
        }
      } else {
        metas = _stmt("SELECT functor, arity FROM qsql_meta").all();
      }
      for (var i = 0; i < metas.length; i++) {
        var m = metas[i];
        if (!_known[m.functor + "/" + m.arity]) continue;
        var tbl = _qsql_tableName(m.functor, m.arity);
        try {
          var rows = _stmt('SELECT _key FROM "' + tbl + '"').all();
          for (var j = 0; j < rows.length; j++) {
            results.push(rows[j]._key);
          }
        } catch(e) { /* table might not exist yet */ }
      }
      return results;
    },

    commit: function() {},

    close: function() {
      _cache = {};
      if (db.close) db.close();
    }
  };
}

// ── Export (dual ESM/CJS) ───────────────────────────────────

if (typeof exports !== "undefined") {
  exports.qsqlAdapter = qsqlAdapter;
  exports._qsql_tableName = _qsql_tableName;
  exports._qsql_argVal = _qsql_argVal;
  exports._qsql_safeName = _qsql_safeName;
}
export { qsqlAdapter, _qsql_tableName, _qsql_argVal, _qsql_safeName };

File: src/qsql.py
# ============================================================
# qsql.py — QSQL: Per-predicate typed SQLite adapter for persist
#
# Zero-impedance bridge: Prolog terms → per-predicate SQLite
# tables with typed argument columns.
#
#   price(aapl, 187.68)  →  table "q$price$2"
#                             _key TEXT PRIMARY KEY
#                             arg0 TEXT   = 'aapl'
#                             arg1 REAL   = 187.68
#
# Atoms → TEXT, numbers → REAL/INTEGER, compounds → JSON TEXT.
# SQLite can index and range-scan individual arguments.
#
# Usage:
#   from qsql import qsql_adapter
#   from persist import persist
#   persist(engine, qsql_adapter("state.db"))
#   persist(engine, qsql_adapter(":memory:"))
# ============================================================

import json
import re


def _safe_name(name):
    return re.sub(r'[^a-zA-Z0-9_]', '_', name)


def _table_name(functor, arity):
    return "q$%s$%d" % (_safe_name(functor), arity)


def _arg_val(arg):
    """Convert a serialized term arg to a native SQLite value."""
    if arg is None:
        return None
    t = arg.get("t")
    if t == "a":
        return arg["n"]
    if t == "n":
        v = arg["v"]
        if isinstance(v, (int, float)):
            return v
        return str(v)  # Decimal, BigFloat → TEXT
    return json.dumps(arg, separators=(',', ':'))


def qsql_adapter(path, parse_fn=None):
    """Create a QSQL adapter for per-predicate typed SQLite storage.

    path      — file path or ":memory:"
    parse_fn  — optional QJSON parse function (default: json.loads)
    """
    import sqlite3
    conn = sqlite3.connect(path)
    conn.execute("PRAGMA journal_mode=WAL")

    _parse = parse_fn or json.loads
    _known = {}  # "functor/arity" → True

    def _ensure_table(functor, arity):
        fa = "%s/%d" % (functor, arity)
        if fa in _known:
            return
        tbl = _table_name(functor, arity)
        cols = ", ".join("arg%d" % i for i in range(arity))
        ddl = 'CREATE TABLE IF NOT EXISTS "%s" (_key TEXT PRIMARY KEY%s)' % (
            tbl, (", " + cols) if cols else "")
        conn.execute(ddl)
        for i in range(arity):
            conn.execute(
                'CREATE INDEX IF NOT EXISTS "ix$%s$%d" ON "%s"(arg%d)' %
                (tbl, i, tbl, i))
        conn.execute(
            "INSERT OR IGNORE INTO qsql_meta VALUES (?, ?)",
            (functor, arity))
        conn.commit()
        _known[fa] = True

    def _setup():
        conn.execute(
            "CREATE TABLE IF NOT EXISTS qsql_meta "
            "(functor TEXT, arity INTEGER, PRIMARY KEY(functor, arity))")
        conn.commit()
        for row in conn.execute("SELECT functor, arity FROM qsql_meta"):
            _known["%s/%d" % (row[0], row[1])] = True

    def _insert(key, functor=None, arity=None):
        if functor is None:
            return
        _ensure_table(functor, arity)
        obj = _parse(key)
        values = [key]
        if obj.get("t") == "c" and "a" in obj:
            for i in range(arity):
                if i < len(obj["a"]):
                    values.append(_arg_val(obj["a"][i]))
                else:
                    values.append(None)
        tbl = _table_name(functor, arity)
        ph = ", ".join("?" for _ in values)
        conn.execute(
            'INSERT OR IGNORE INTO "%s" VALUES (%s)' % (tbl, ph),
            tuple(values))

    def _remove(key):
        try:
            obj = _parse(key)
        except (ValueError, TypeError):
            return
        if obj.get("t") == "c":
            functor = obj["f"]
            arity = len(obj.get("a", []))
        elif obj.get("t") == "a":
            functor = obj["n"]
            arity = 0
        else:
            return
        fa = "%s/%d" % (functor, arity)
        if fa not in _known:
            return
        tbl = _table_name(functor, arity)
        conn.execute('DELETE FROM "%s" WHERE _key = ?' % tbl, (key,))

    def _all(predicates=None):
        results = []
        if predicates:
            metas = []
            for pred in predicates:
                parts = pred.split("/")
                metas.append((parts[0], int(parts[1])))
        else:
            metas = conn.execute(
                "SELECT functor, arity FROM qsql_meta").fetchall()
        for functor, arity in metas:
            fa = "%s/%d" % (functor, arity)
            if fa not in _known:
                continue
            tbl = _table_name(functor, arity)
            try:
                rows = conn.execute(
                    'SELECT _key FROM "%s"' % tbl).fetchall()
                for row in rows:
                    results.append(row[0])
            except Exception:
                pass
        return results

    return {
        "setup": _setup,
        "insert": _insert,
        "remove": _remove,
        "all": _all,
        "commit": lambda: conn.commit(),
        "close": lambda: conn.close(),
    }
